streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -854,13 +881,48 @@ public void onAssignment(final Assignment assignment) {
                     usedSubscriptionMetadataVersion = leaderSupportedVersion;
                 }
                 processVersionFourAssignment(info, partitions, activeTasks, topicToPartitionInfo);
-                partitionsByHost = info.partitionsByHost();
+                partitionsByHost = info.getPartitionsByHost();
+                break;
+            case VERSION_FIVE:
+                if (leaderSupportedVersion > usedSubscriptionMetadataVersion) {
+                    log.info("Sent a version {} subscription and group leader's latest supported version is {}. " +
+                                    "Upgrading subscription metadata version to {} for next rebalance.",
+                            usedSubscriptionMetadataVersion,
+                            leaderSupportedVersion,
+                            leaderSupportedVersion);
+                    usedSubscriptionMetadataVersion = leaderSupportedVersion;
+                }
+                processVersionFiveAssignment(info, partitions, activeTasks, topicToPartitionInfo);
+                tasksByHosts = info.tasksByHost();
+
+                final Map<Integer, InternalTopologyBuilder.TopicsInfo> topicGroups = taskManager.builder().topicGroups();
+                final Map<Integer, Set<String>> sourceTopicsByGroup = new HashMap<>();
+                for (final Map.Entry<Integer, InternalTopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
+                    sourceTopicsByGroup.put(entry.getKey(), entry.getValue().sourceTopics);
+                }
+
+                final Cluster fullMetadata = Cluster.empty().withPartitions(topicToPartitionInfo);
+
+                partitionsByHost = new HashMap<>();
+                if (userEndPoint != null && !(userEndPoint.isEmpty()) && receivedAssignmentMetadataVersion >= 5) {
+                    final Map<TaskId, Set<TopicPartition>> partitionsForTask = partitionGrouper.partitionGroups(sourceTopicsByGroup, fullMetadata);
+
+                    for (final Map.Entry<HostInfo, Set<TaskId>> entry : tasksByHosts.entrySet()) {
+                        final Set<TopicPartition> topicPartitions = new HashSet<>();
+                        for (final TaskId taskId : entry.getValue()) {
+                            topicPartitions.addAll(partitionsForTask.get(taskId));
```

Comment:
@guozhangwang I don't quite follow your note on handling the case where a single task processes >1 partition from the same topic. I don't see any keying based on topics being done. where do you think the issue will surface?